cpp/src/round/round.cu

```diff
@@ -0,0 +1,131 @@
+
+namespace detail {
+
+struct round_fn {
+  template <typename T, typename... Args>
+  std::enable_if_t<not cudf::is_numeric<T>(), std::unique_ptr<column>> operator()(Args&&... args)
+  {
+    CUDF_FAIL("Type not currenlty support for cudf::round");
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::round_option round,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result   = cudf::make_fixed_width_column(input.type(), input.size());
+    auto out_view = result->mutable_view();
+    auto const n  = static_cast<int32_t>(std::pow(10, std::abs(decimal_places)));
+
+    if (decimal_places == 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [] __device__(T e) -> T { return ::round(e); });
+    else if (decimal_places > 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return ::round(e * n) / n; });
+    else  // decimal_places < 0
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return ::round(e / n) * n; });
+
+    return result;
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_integral<T>::value, std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::round_option round,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result = cudf::make_fixed_width_column(input.type(), input.size());
+
+    // only need to handle the case where decimal_places is < zero
+    // integers by definition have no fractional part, so result of "rounding" is a no-op
+    if (decimal_places < 0) {
+      auto out_view = result->mutable_view();
+      auto const n  = static_cast<int64_t>(std::pow(10, -decimal_places));
+      auto const m  = n / 10;
+
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n, m] __device__(T e) -> T {
+                          auto const rounding_digit = static_cast<int64_t>(fabs((e / m) % 10));
+                          auto const digits         = e / n;
+                          auto const adjust         = e > 0 ? 1 : -1;
+                          return rounding_digit < 5 ? digits * n : (digits + adjust) * n;
+                        });
```

Comment:
Why can't this just be (e / n) * n?  round(1576, -2) --> (1576 / 100) * 100 --> 1500?

Reply:
`(e / n) * n` would do integer division and fail for the cases where you need to round up. Recall that I previously had:
```cpp
return std::roundf(1.0 * e / n) * n; 
```
but revans2 pointed out:
> Is this going to do the right thing if we have an INT64? If I have any int value that does not fit in the 23 bit fraction of a float combined with a decimal_places close to 0 (like -1) we would end up with a wrong answer because we lost information in the implicit cast to a float.

So I addd the `Int64AtBoundary` test, found that he was indeed correct - and added the code above to work.

Reply:
Duh.